wasm/src/main/java/com/dylibso/chicory/wasm/io/WasmInputStream.java

```diff
@@ -0,0 +1,641 @@
+     * @throws WasmParseException if the format of the stream is invalid
+     * @throws WasmIOException if an underlying I/O error occurred
+     * @throws WasmEOFException if the end-of-file was reached unexpectedly
+     */
+    public List<ValueType> typeVec() throws WasmIOException {
+        int cnt = u31();
+        switch (cnt) {
+            case 0:
+                return List.of();
+            case 1:
+                return List.of(type());
+            case 2:
+                return List.of(type(), type());
+            default:
+                ValueType[] types = new ValueType[cnt];
+                for (int i = 0; i < cnt; i++) {
+                    types[i] = type();
+                }
+                return List.of(types);
+        }
+    }
+
+    /**
+     * Read a value type from the stream.
+     *
+     * @return the value type (not {@code null})
+     * @throws WasmParseException if the format of the stream is invalid
+     * @throws WasmIOException if an underlying I/O error occurred
+     * @throws WasmEOFException if the end-of-file was reached unexpectedly
+     */
+    public ValueType type() throws WasmIOException {
+        try {
+            return ValueType.forId(rawByte());
+        } catch (IllegalArgumentException e) {
+            throw badType(e);
+        }
+    }
+
+    /**
+     * Read a WASM reference type from the stream.
+     *
+     * @return the reference type (not {@code null})
+     * @throws WasmParseException if the format of the stream is invalid
+     * @throws WasmIOException if an underlying I/O error occurred
+     * @throws WasmEOFException if the end-of-file was reached unexpectedly
+     */
+    public ValueType refType() throws WasmIOException {
+        try {
+            return ValueType.refTypeForId(rawByte());
+        } catch (IllegalArgumentException e) {
+            throw badType(e);
+        }
+    }
+
+    /**
+     * Read a WASM mutability type from the stream.
+     *
+     * @return the mutability type (not {@code null})
+     * @throws WasmParseException if the format of the stream is invalid
+     * @throws WasmIOException if an underlying I/O error occurred
+     * @throws WasmEOFException if the end-of-file was reached unexpectedly
+     */
+    public MutabilityType mut() throws WasmIOException {
+        try {
+            return MutabilityType.forId(rawByte());
+        } catch (IllegalArgumentException e) {
+            throw badType(e);
+        }
+    }
+
+    private static WasmParseException badType(Throwable cause) {
+        return new WasmParseException("Invalid type", cause);
+    }
+
+    /**
+     * Read a ULEB value returning up to 64 significant bits, with range validation.
+     *
+     * @param bits the number of significant bits allowed
+     * @return the value
+     * @throws WasmIOException if an underlying I/O error occurs or the value exceeds the allowed range
+     */
+    private long uleb(int bits) throws WasmIOException {
```

Comment:
We already have Leb implementations in `com.dylibso.chicory.wasm.Encoding`, I think we should consolidate on one.

Reply:
That class will no longer be used after this change. A follow-up PR will remove it.

Reply:
Any challenge including the change in this PR in the spirit of keeping things more "atomic"?

Reply:
Sure, I can move that change.

Reply:
Looks like I have to do some commit log surgery to get the whole thing. One moment.

Reply:
Now it should be OK.